pubnub-chat-impl/src/commonMain/kotlin/com/pubnub/chat/internal/MessageDraftImpl.kt

```diff
@@ -0,0 +1,387 @@
+                    channels.map { channel ->
+                        SuggestedMention(
+                            matchResult.matchStart,
+                            matchResult.matchString,
+                            channel.name ?: channel.id,
+                            MentionTarget.Channel(channel.id)
+                        )
+                    }
+                }
+            }.awaitAll()
+        return awaitAll(getSuggestedUsersFuture, getSuggestedChannelsFuture).then {
+            it.first.flatten() + it.second.flatten()
+        }
+    }
+
+    private fun triggerTypingIndicator() {
+        if (isTypingIndicatorTriggered) {
+            if (messageText.isNotEmpty()) {
+                channel.startTyping()
+            } else {
+                channel.stopTyping()
+            }
+        }
+    }
+
+//    private val MatchResult.matchStart get() = groups[1]!!.range.first // not supported in JS
+    private val MatchResult.matchStart get() = range.first
+    private val MatchResult.matchString get() = groups[1]!!.value
+
+    override fun insertText(offset: Int, text: String) {
+        insertTextInternal(offset, text)
+        triggerTypingIndicator()
+        fireMessageElementsChanged()
+    }
+
+    private fun insertTextInternal(offset: Int, text: String) {
+        messageText.insert(offset, text)
+        val toRemove = mutableSetOf<Mention>()
+        mentions.forEach { mention ->
+            when {
+                offset > mention.start && offset < mention.endExclusive -> {
+                    toRemove += mention
+                }
+                offset <= mention.start -> {
+                    mention.start += text.length
+                }
+            }
+        }
+        mentions.removeAll(toRemove)
+    }
+
+    override fun removeText(offset: Int, length: Int) {
+        removeTextInternal(offset, length)
+        triggerTypingIndicator()
+        fireMessageElementsChanged()
+    }
+
+    private fun removeTextInternal(offset: Int, length: Int) {
+        messageText.deleteRange(offset, offset + length)
+        val toRemove = mutableSetOf<Mention>()
+        mentions.forEach { mention ->
+            val removalEnd = offset + length
+            if (offset <= mention.endExclusive && removalEnd > mention.start) {
+                toRemove += mention
+            }
+            if (offset < mention.start) {
+                val numCharactersBefore = min(length, mention.start - offset)
+                mention.start -= numCharactersBefore
+            }
+        }
+        mentions.removeAll(toRemove)
+    }
+
+    override fun insertSuggestedMention(mention: SuggestedMention, text: String) {
+        if (messageText.substring(mention.start, mention.start + mention.replaceFrom.length) != mention.replaceFrom) {
+            throw PubNubException("This mention suggestion is no longer valid - the message draft text has been changed.")
```

Comment:
Could error message be moved to PubNubErrorMessage ?

Reply:
done